components/system-broker/pkg/http/secured_transport.go

```diff
@@ -19,64 +19,59 @@ package http
 import (
 	"context"
 	"net/http"
-	"sync"
-	"time"
 
+	"github.com/kyma-incubator/compass/components/director/pkg/log"
 	"github.com/pkg/errors"
 )
 
 //go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . TokenProvider
 type TokenProvider interface {
+	Name() string
+	Matches(ctx context.Context) bool
 	GetAuthorizationToken(ctx context.Context) (Token, error)
 }
 
 type SecuredTransport struct {
-	timeout       time.Duration
-	roundTripper  HTTPRoundTripper
-	tokenProvider TokenProvider
-	lock          sync.RWMutex
-
-	token Token
+	roundTripper   HTTPRoundTripper
+	tokenProviders []TokenProvider
 }
 
-func NewSecuredTransport(timeout time.Duration, roundTripper HTTPRoundTripper, provider TokenProvider) *SecuredTransport {
+func NewSecuredTransport(roundTripper HTTPRoundTripper, providers ...TokenProvider) *SecuredTransport {
 	return &SecuredTransport{
-		timeout:       timeout,
-		roundTripper:  roundTripper,
-		tokenProvider: provider,
-		lock:          sync.RWMutex{},
+		roundTripper:   roundTripper,
+		tokenProviders: providers,
 	}
 }
 
 func (c *SecuredTransport) RoundTrip(request *http.Request) (*http.Response, error) {
-	if err := c.refreshToken(request.Context()); err != nil {
+	logger := log.C(request.Context())
+
+	token, err := c.getToken(request.Context())
```

Comment:
Why do you need to put "some configured headers" in the context in order to use them later for the token provider as you have here the request. Can't the token provider receive the whole request object and get the headers from there?
May be there is something I don't see for the need of the context, but in my opinion it would be more intuitive if you access the headers directly.

Reply:
As discussed offline this approach is needed because of the Bind API of the broker framework because it does not have a http.Request